### PR TITLE
Update automated-tests.md

### PR DESCRIPTION
**Pull request by 3hhh**
Discard instrusctions for Qubes 3.2 & make it actually work.

### CODE REVIEW

File: developer/debugging/automated-tests.md [hunk 1]
```diff
@@ -17,23 +17,31 @@ Starting with Qubes R3 we use [python unittest][unittest] to perform automatic t
...
+Since the tests were written with this expectation, all the VMs with a name starting with `test-` on the installation are removed during the process, and all the tests are recklessly started from dom0, even when testing VM components.
 
-Our test runner runs mostly the same as the standard one, but it has some nice additional features like color output and not needing the "qubes.test" prefix. 
-It also has the ability to run lone selected template tests.
+Most of the tests are stored in the [core-admin repository](https://github.com/QubesOS/qubes-core-admin/tree/master/qubes/tests) in the `qubes/tests` directory. 
 
-You can use `python3 -m qubes.tests.run -h` to get usage information:
+To run tests, you need to do the following steps:
+1. `git clone https://github.com/QubesOS/qubes-core-qrexec` and put it to `/root/` in dom0 on your test machine.
+2. `git clone https://github.com/QubesOS/qubes-core-admin` and put it to `/root/` in dom0 on your test machine.
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:30]: You don't need to download it. Both are already installed in dom0. And even worse - if you for example download "master" branch into Qubes 4.0 system, it won't be compatible at all.

File: developer/debugging/automated-tests.md [hunk 2]
```diff
@@ -17,23 +17,31 @@ Starting with Qubes R3 we use [python unittest][unittest] to perform automatic t
...
+3. Inside `dom0` on your test machine, create a script at `/root/qubes-core-qrexec/run`:
+    ```{.bash}
+    #!/bin/bash
+    [[ "$(whoami)" == "root" ]] || { >&2 echo "This script must be run as root." ; exit 1 ; }
+    systemctl stop qubesd
+    export PYTHONPATH=/root/qubes-core-qrexec
+    ./run-tests -v $(python3 -m qubes.tests.run -l | grep "$@")
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:37]: The "run-tests" script is a helper script for running tests directly from the repository, not necessary in dom0. This is just for selected unit tests that actually can run outside of dom0.
If you run them in dom0, use `python3 -m qubes.tests.run` directly, as you can see in instructions for Qubes 4.0.
If you like, you can still provide it a list of tests to run (as shows in an example below).

File: developer/debugging/automated-tests.md [hunk 3]
```diff
@@ -17,23 +17,31 @@ Starting with Qubes R3 we use [python unittest][unittest] to perform automatic t
...
-You can use `python3 -m qubes.tests.run -h` to get usage information:
+To run tests, you need to do the following steps:
+1. `git clone https://github.com/QubesOS/qubes-core-qrexec` and put it to `/root/` in dom0 on your test machine.
+2. `git clone https://github.com/QubesOS/qubes-core-admin` and put it to `/root/` in dom0 on your test machine.
+3. Inside `dom0` on your test machine, create a script at `/root/qubes-core-qrexec/run`:
+    ```{.bash}
+    #!/bin/bash
+    [[ "$(whoami)" == "root" ]] || { >&2 echo "This script must be run as root." ; exit 1 ; }
+    systemctl stop qubesd
+    export PYTHONPATH=/root/qubes-core-qrexec
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:36]: Again, not necessary and quite likely harmful (if you mess up branches). Qrexec is already in dom0.

File: developer/debugging/automated-tests.md [hunk 4]
```diff
@@ -73,9 +81,9 @@ You can use `python3 -m qubes.tests.run -h` to get usage information:
     MODULE+"/"+CLASS+"/"+FUNCTION. MODULE should omit initial "qubes.tests.".
     Example: basic/TC_00_Basic/test_000_create
 
-For instance, to run only the tests for the fedora-21 template, you can use the `-l` option, then filter the list:
+For instance, to find only the tests for the fedora-21 template, you can use the `-l` option, then filter the list:
 
-    [user@dom0 ~]$ python3 -m qubes.tests.run -l | grep fedora-21
+    [user@dom0 qubes-core-admin]$ python3 -m qubes.tests.run -l | grep fedora-21
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:86]: The directory is irrelevant. This should use installed `qubes` package.

**Comment by 3hhh** [comment]
Maybe there was a general misunderstanding there:

I expected the doc page to tell me how to test code from my branch.

Apparently however the page was meant to tell me how to test the code from the branch that Qubes OS installed?

So if I understand you correctly, I can test my branch inside a VM as well?

I guess it would be good to re-work the page in total with 2 sections then: One about the "test your own branch inside a VM" and one about the "test Qubes OS installed" inside dom0? For most devs the former is more relevant...

**Comment by marmarek** [comment]
> Apparently however the page was meant to tell me how to test the code from the branch that Qubes OS installed?

Yes, the integration tests documented on that page are testing the whole system, not only core-admin repository. This requires the things you want to test to be installed in the test system.

> So if I understand you correctly, I can test my branch inside a VM as well?

Yes, `run-tests` script in core-admin can be used for that. Indeed currently it doesn't take care about qrexec module dependency that was added not long ago. Personally I run `PYTHONPATH=../core-qrexec ./run-tests` (it works since I have core-admin side by side with core-qrexec, as qubes-builder does). Perhaps this should be integrated into `run-tests` script directly.

> I guess it would be good to re-work the page in total with 2 sections then: One about the "test your own branch inside a VM" and one about the "test Qubes OS installed" inside dom0? For most devs the former is more relevant...

Yes, that would make sense. Although keep in mind the ability to test various parts outside of dom0 is limited. In case of storage drivers - you can test creating, removing, resizing volumes (unit tests), but you can not test if VM can start with it connected (integration tests). Both approaches are important, so including unit tests documentation is a good idea.

**Comment by 3hhh** [comment]
I updated the PR as per our discussion.

File: developer/debugging/automated-tests.md [hunk 5]
```diff
@@ -96,28 +109,46 @@ For instance, to run only the tests for the fedora-21 template, you can use the
...
+Many unit tests will also work inside a VM. However all of the tests requiring a dedicated VM to be run (mostly the integration tests) will be skipped.
+
+As minimal requirements for testing, you'll have to clone the [qubes-core-admin](https://github.com/QubesOS/qubes-core-admin) and [qubes-core-qrexec](https://github.com/QubesOS/qubes-core-qrexec) repositories.
+
+The below example however will assume that you set up a build environment as described in the [Qubes Builder documentation](/doc/qubes-builder/).
+
+Assuming you cloned the `qubes-builder` repository to your home directory inside a fedora VM, you can use the following commands to run the unit tests:
+```{.bash}
+sudo dnf install python3-pip lvm2 python35 python3-virtualenv
+virtualenv -p /usr/bin/python35 python35
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:136]: This shouldn't be strictly necessary (code written for python 3.5 should work on newer python too), but indeed can improve test results (detect things that do work on python 3.7 for example, but will break on python 3.5).

But more importantly - the version to test on depends on Qubes version (what runs in dom0):
 - Qubes 4.0 (`release4.0` branch): python >=3.5
 - Qubes 4.1 (`master` branch): python >=3.7

File: developer/debugging/automated-tests.md [hunk 6]
```diff
@@ -96,28 +109,46 @@ For instance, to run only the tests for the fedora-21 template, you can use the
...
 
-On Qubes 4.0 tests are also compatible with nose2 test runner, so you can use this instead:
 
-    sudo systemctl stop qubesd; sudo -E nose2 -v --plugin nose2.plugins.loader.loadtests qubes.tests; sudo systemctl start qubesd
+### Unit testing inside a VM
 
-This may be especially useful together with various nose2 plugins to store tests results (for example `nose2.plugins.junitxml`), to ease presenting results. This is what we use on [OpenQA].
+Many unit tests will also work inside a VM. However all of the tests requiring a dedicated VM to be run (mostly the integration tests) will be skipped.
+
+As minimal requirements for testing, you'll have to clone the [qubes-core-admin](https://github.com/QubesOS/qubes-core-admin) and [qubes-core-qrexec](https://github.com/QubesOS/qubes-core-qrexec) repositories.
```
Comment (marmarek) [inline developer/debugging/automated-tests.md:129]: To be clear - you need qubes-core-admin if you want to run unit tests of qubes-core-admin. And currently indeed unit tests there require core-qrexec too.
But there are unit tests in other repositories too for example qubes-core-admin-client, desktop-linux-common, desktop-linux-manager and some more. You need just the respective repository for running unit tests from them.

The generic sentence "Most of the tests are stored in the core-admin repository" applies to integration tests - which this page was about previously.

**Pull request merged**